### ISSUE TITLE
LSIF Go Indexing fails after 3.17 > 3.21 upgrade

### ISSUE THREAD

**Issue by christinelovett** [sourcegraph/sourcegraph#16030]
3.21.2
Kubernetes 
Go repo
src-cli version 3.22.2

After upgrading from 3.17 to 3.21.2 LSIF Go indexing is failing. src-cli is on version 3.22.2. Worker counts and resources have been increased but didn't resolve the issue. 

No logs on worker container. 

precise-code-intel-bundle-manager logs:

kubectl logs --tail 500 precise-code-intel-bundle-manager-678bf99578-lm8hq
t=2020-11-17T18:42:34+0000 lvl=eror msg=store.HardDeleteUploadByID error="pq: deadlock detected" count=1.000 elapsed=1.032

t=2020-11-17T18:42:34+0000 lvl=eror msg="Failed to hard deleted upload records" error="pq: deadlock detected"

t=2020-11-18T15:40:26+0000 lvl=eror msg="Failed to write payload to client" err="write tcp 172.168.145.70:3187->172.168.139.198:45284: write: connection reset by peer"

2020/11/18 18:55:58 received error during background config update, err: unable to fetch new configuration: Post "http://sourcegraph-frontend-internal/.internal/configuration": EOF

2020/11/18 19:31:22 received error during background config update, err: unable to fetch new configuration: Post "http://sourcegraph-frontend-internal/.internal/configuration": EOF

2020/11/18 21:52:50 received error during background config update, err: unable to fetch new configuration: Post "http://sourcegraph-frontend-internal/.internal/configuration": read tcp 172.168.145.70:59870->172.20.183.112:80: read: connection reset by peer

t=2020-11-19T07:37:08+0000 lvl=eror msg=store.HardDeleteUploadByID error="pq: deadlock detected" count=1.000 elapsed=1.002

t=2020-11-19T07:37:08+0000 lvl=eror msg="Failed to hard deleted upload records" error="pq: deadlock detected

https://app.hubspot.com/contacts/2762526/company/464956351

**Comment by efritz** [comment]
We've pin-poined the issue here! After 3.19, we changed the way we calculated the set of indexes can serve code intelligence for a path at a commit. Before, we would backfill any missing commits on each query and do a recursive scan on the commits table to find the closest set of indexes. After, we denormalize this data to make the lookup O(1) at the expense of a periodic refresh.

This runs in the worker, which is crashing at regular intervals at this customer due to OOM. We need to reduce the memory pressure created by [this function](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@d0cdec9fa9385222e5e4545de3e427e4e3c13f00/-/blob/enterprise/internal/codeintel/store/commits.go#L102:17).

Increasing the memory of the worker instance can be a workaround until we can patch the performance.
